### PR TITLE
Fix tick bug when expanding electric potential to full 2pi

### PR DESCRIPTION
**Pull request by SebastianRuffert**
This should solve issue #253

### CODE REVIEW

**Comment by lmh91** [comment]
The field calculations seems to work but I can't plot the electric potential and electric field in r-phi:
```julia
plot(sim.electric_potential, z= 0)
```
Could you please have a look?

**Comment by fhagemann** [comment]
Yes, the ticks should be ordered, and ideally, the first tick of a `DiscreteAxis` should also be the left boundary of the respective `interval`, i.e.
```julia
g = Grid(sim)
g.φ.ticks[1] == g.φ.interval.left
```

File: src/ScalarPotentials/ScalarPotential.jl [hunk 1]
```diff
@@ -48,14 +48,15 @@ function get_2π_potential(sp::ScalarPotential{T, 3, Cylindrical}, axφ::Discret
...
             new_pot[:, idx, :] = sp[:, il, :]
         end
     end
-    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks )
+    P = sortperm(new_ticks)
+    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks[P] )
     new_axes = (sp.grid[1], new_axφ, sp.grid[3])
     new_grid = Grid{AT, 3, Cylindrical, typeof(new_axes)}( new_axes )
-    ScalarPotential(sp, new_pot, new_grid)
+    ScalarPotential(sp, new_pot[:,P,:], new_grid)
```
Comment (fhagemann) [inline src/ScalarPotentials/ScalarPotential.jl:59]: Also here, `g.φ.ticks[1]` is not necessarily equal to `g.φ.interval.left`.
Please compose an example where `g.φ.ticks` does not include `0` or multiples of `2pi` (probably a 2D symmetric detector from `phi: [17°,17°]` or something like this and check whether the expansion of the potential to full `2pi` affects the calculation, plotting, etc.
It would also be good to add such a case with a non-zero left `phi` entry for the `grid` in the config file to the tests.

**Review by fhagemann — CHANGES_REQUESTED** [review]

File: src/ScalarPotentials/ScalarPotential.jl (hunk 1 above)
Comment (SebastianRuffert) [reply src/ScalarPotentials/ScalarPotential.jl]: I already added such a geometry and will write tests. Though when using a 2D geometry we only test `extend_2D_to_3D_by_n_points`. We are not testing `get_2π_potential` for periodic boundary conditions. What is not yet implemented is `get_2π_potential` for reflecting boundaries, which should also be tested.

File: src/ScalarPotentials/ScalarPotential.jl [hunk 2]
```diff
@@ -81,24 +83,26 @@ end
...
     Δφ::AT = 2π / n_points_in_φ
     for i in 1:n_points_in_φ
-        new_ticks[i] = (i - 1) * Δφ
+        new_ticks[i] = mod((i - 1) * Δφ + sp.grid[2][1],T(2π))
         new_pot[:, i, :] = sp[:, 1, :]
     end
-    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks )
+    P = sortperm(new_ticks)
+    new_int::Interval{:closed, :open, AT} = Interval{:closed, :open, AT}(new_ticks[P][1], new_ticks[P][end])
+    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks[P] )
```
Comment (fhagemann) [inline src/ScalarPotentials/ScalarPotential.jl:96]: Then, this is also not needed (or needs to be modified according to my first comment.

File: src/ScalarPotentials/ScalarPotential.jl [hunk 3]
```diff
@@ -41,21 +41,23 @@ function get_2π_potential(sp::ScalarPotential{T, 3, Cylindrical}, axφ::Discret
...
             idx::Int = il + (idx_n - 1) * l
-            new_ticks[idx] = (idx_n - 1) * Δφ + axφ.ticks[il]
+            new_ticks[idx] = mod((idx_n - 1) * Δφ + axφ.ticks[il], T(2π))
             new_pot[:, idx, :] = sp[:, il, :]
         end
     end
-    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks )
+    P = sortperm(new_ticks)
+    new_int::Interval{:closed, :open, AT} = Interval{:closed, :open, AT}(new_ticks[P][1], new_ticks[P][end])
+    new_axφ = DiscreteAxis{AT, :periodic, :periodic}( new_int, new_ticks[P] )
```
Comment (fhagemann) [inline src/ScalarPotentials/ScalarPotential.jl:57]: I do not think that this is what we want...

The endpoints of the `Interval` should not necessarily be the first and last ticks of `new_ticks[P]`, especially when the `Interval` is `:open` on the right.

In this function (`get_2π_potential`), we want to end up with a `2π` potential,and in my personal opinion the `new_int` should also be a `2π`-`Interval`, i.e. where the `:closed`-`:open`-`Interval` ranges from `[0,2π[` (or `[x,x+2π[` but then again there might be values outside of the `[0,2π[` range...)

File: src/ScalarPotentials/ScalarPotential.jl [hunk 4]
```diff
@@ -81,24 +83,26 @@ end
...
 function extend_2D_to_3D_by_n_points(sp::ScalarPotential{T, 3, Cylindrical}, axφ::DiscreteAxis{AT, :reflecting, :reflecting}, int::Interval{:closed, :closed, AT},
         n_points_in_φ::Int ) where {T, AT}
-    new_int::Interval{:closed, :open, AT} = Interval{:closed, :open, AT}(0, 2π)
+    #new_int::Interval{:closed, :open, AT} = Interval{:closed, :open, AT}(0, 2π)
     new_ticks::Vector{AT} = Vector{AT}(undef, n_points_in_φ)
     new_pot = Array{eltype(sp.data), 3}(undef, size(sp, 1), n_points_in_φ, size(sp, 3))
     Δφ::AT = 2π / n_points_in_φ
     for i in 1:n_points_in_φ
-        new_ticks[i] = (i - 1) * Δφ
+        new_ticks[i] = mod((i - 1) * Δφ + sp.grid[2][1],T(2π))
```
Comment (fhagemann) [inline src/ScalarPotentials/ScalarPotential.jl:91]: If we start from a `2D` potential, I do not think that we need this modification. We can just use the previous code and have the electric potential range from `[0,2π[`. (Also, I do not see why we would need to include the original `2D` tick value in the `new_ticks` for the `3D` potential here, as everything will be `φ`-symmetric).

File: examples/example_config_files/cone_config.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+              from: 1.0
+              to: 10.0    
+          h: 10.
+          origin:
+            z: 5
+    contacts:
+      - name: "p contact"
+        id: 1
+        material: "HPGe"
+        potential: 4000.
```
Comment (fhagemann) [inline examples/example_config_files/cone_config.yaml:48]: Minor comment: p+ contacts usually have a negative potential applied

File: examples/example_config_files/cone_config.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      from: -5
+      to: 15
+      boundaries:
+        left: "inf"
+        right: "inf"
+medium: "vacuum"
+detectors:
+  - semiconductor:
+      material: "HPGe"
+      bulk_type: "p"
```
Comment (fhagemann) [inline examples/example_config_files/cone_config.yaml:27]: `bulk_type`, at least to my knowledge, should be ignored (?)

File: src/ScalarPotentials/ScalarPotential.jl (hunk 4 above)
Comment (SebastianRuffert) [reply src/ScalarPotentials/ScalarPotential.jl:91]: That is correct, I did this mainly for consistency reasons.

**Comment by fhagemann** [comment]
Why was this closed?

**Comment by lmh91** [comment]
@fhagemann What is the status of this? 
It looks good to me. Can this be merged?

**Comment by fhagemann** [comment]
I did not yet look into this in detail, might get to this next week. What I see is that we added new config files without corresponding test. It might make sense to add tests for a phi interval not starting at 0 to make sure that everything runs smoothly before merging this into main.

**Comment by fhagemann** [comment]
If I have a look at the new config file introduced in this PR (`ConeSym`), the point types look non-2π-symmetric for some `z`, e.g.

```julia
using SolidStateDetectors, Unitful, Plots
sim = Simulation(SSD_examples[:ConeSym])
calculate_electric_potential!(sim)
plot(sim.point_types, z = 0.009)
```
image

Not sure exactly what the problem is here (maybe missing important ticks?) but this is something to look into before merging.

**Comment by fhagemann** [comment]
> image
> 
> Not sure exactly what the problem is here (maybe missing important ticks?) but this is something to look into before merging.

This happens to be due to rounding errors 
(one tick at `0.00575` and one at `0.00625`, and the intersection is at `0.006` or `0.0059999`..)

File: src/ScalarPotentials/ScalarPotential.jl [hunk 7]
```diff
@@ -48,14 +48,35 @@ function get_2π_potential(sp::ScalarPotential{T, 3, Cylindrical}, axφ::Discret
...
+        new_axes = (sp.grid[1], new_axφ, sp.grid[3])
+        new_grid = Grid{AT, 3, Cylindrical, typeof(new_axes)}( new_axes )
+        ScalarPotential(sp, new_pot[:,P,:], new_grid)
+    else
+        new_ticks_new_zero::Vector{AT} = Vector{AT}(undef, l * n + 1)
+        new_ticks_new_zero[1] = T(0)
+        new_ticks_new_zero[2:end] = new_ticks[P]
+        new_pot_new_zero = Array{eltype(sp.data), 3}(undef, size(sp, 1), l * n + 1, size(sp, 3))
+        new_pot_new_zero[:,1,:] = (eltype(sp.data) <: AbstractFloat) ? sp[:,1,:] + 
+                        (sp[:,end,:] - sp[:,1,:]) * (T(0) - new_ticks_new_zero[2]) / (T(2π) - T(new_ticks_new_zero[end]) * n + T(new_ticks_new_zero[2])) :
```
Comment (fhagemann) [inline src/ScalarPotentials/ScalarPotential.jl:67]: I feel like this formula is wrong and it should contain something like `Δφ * n` rather than `T(new_ticks_new_zero[end]) * n`.. I will do the calculation and see if this needs to be adjusted.

Comment (fhagemann) [reply src/ScalarPotentials/ScalarPotential.jl:67]: The permutation of the new potential based on the permutation of the new ticks is missing here, I will push some changes to fix this..
